### ISSUE TITLE
numpy delete where is giving wrong result i.e. always deleting the first element if there is atleast one True

### ISSUE THREAD

**Issue by ragorepo** [numpy/numpy#18412]
```
xyz =np.array( [[[612.        ,  0.8679449]],
                [[612.        ,  0.7679449]],

 [[206.,          0.338741 ]],

 [[62.,          2.338741 ]]])
xyx = np.copy(xyz)
np.delete(xyx, np.where([[False], [True],[True],[ True]]), axis=0)

```
Output:
array([], shape=(0, 1, 2), dtype=float64)

```
`xyz =np.array( [[[612.        ,  0.8679449]],
                [[612.        ,  0.7679449]],

 [[206.,          0.338741 ]],

 [[62.,          2.338741 ]]])
xyx = np.copy(xyz)
np.delete(xyx, np.where([[False], [False],[False],[ True]]), axis=0)`
```

output
array([[[6.120000e+02, 7.679449e-01]],

       [[2.060000e+02, 3.387410e-01]]])

```
xyz =np.array( [[[612.        ,  0.8679449]],
                [[612.        ,  0.7679449]],

 [[206.,          0.338741 ]],

 [[62.,          2.338741 ]]])
xyx = np.copy(xyz)
np.delete(xyx, np.where([[False], [False],[False],[ False]]), axis=0)
```
but this works fine as output below.
array([[[6.120000e+02, 8.679449e-01]],

       [[6.120000e+02, 7.679449e-01]],

       [[2.060000e+02, 3.387410e-01]],

       [[6.200000e+01, 2.338741e+00]]])

Issue is observed- if there is atleast one True in where condition, then it is also deleting the first row.

**Comment by eric-wieser** [comment]
Can you provide information about which version of numpy you're using, as was requested by the issue template?

**Comment by ragorepo** [comment]
> Can you provide information about which version of numpy you're using, as was requested by the issue template?

It is 1.20.1

**Comment by seberg** [comment]
`np.delete` does not understand multiple dimensions, your `np.where` call means that you effectively call:
```
np.delete(arr, np.where(bools), axis=0)
```
translating to:
```
np.delete(arr, np.asarray(np.where(bools)), axis=0)
```
And since you specify `axis=0`, `np.delete` assumes that all elements in that array should be deleted, which is the same as:
```
np.delete(arr, np.asarray(np.where(bools)).ravel(), axis=0)
```
Note that:
```
>>> np.array(np.where([[False], [True],[True],[ True]])).ravel()
array([1, 2, 3, 0, 0, 0])
```

Now, maybe `np.delete` should just refuse a multi-dimensional input here...

---

I would suggest you use indexing instead, which is probably nicer anyway:
```
mask = np.array([[False], [False],[False],[ True]])
result = arr[~mask]
```

**Comment by TechDevi** [comment]
I used below code and it is working perfectly fine.

**Comment by Ra-works** [comment]
Thanks for the reply. I had proceeded with alternative approach but is this
delete failure  bug or not?

**Comment by Ra-works** [comment]
> I used below code and it is working perfectly fine. I am new to Python, so not sure if you have any specific reason to use a multi-dimensional array in 'np.where'.
> 
> xyz =np.array( [[[612. , 0.8679449]],
> [[612. , 0.7679449]],
> 
> [[206., 0.338741 ]],
> 
> [[62., 2.338741 ]]])
> xyx = np.copy(xyz)
> np.delete(xyx, np.where([False,False,True, True]), axis=0)

Yes - We have to preserve the dimensions while doing operations..

**Comment by Ra-works** [comment]
> `np.delete` does not understand multiple dimensions, your `np.where` call means that you effectively call:
> 
> ```
> np.delete(arr, np.where(bools), axis=0)
> ```
> 
> translating to:
> 
> ```
> np.delete(arr, np.asarray(np.where(bools)), axis=0)
> ```
> 
> And since you specify `axis=0`, `np.delete` assumes that all elements in that array should be deleted, which is the same as:
> 
> ```
> np.delete(arr, np.asarray(np.where(bools)).ravel(), axis=0)
> ```
> 
> Note that:
> 
> ```
> >>> np.array(np.where([[False], [True],[True],[ True]])).ravel()
> array([1, 2, 3, 0, 0, 0])
> ```
> 
> Now, maybe `np.delete` should just refuse a multi-dimensional input here...
> 
> I would suggest you use indexing instead, which is probably nicer anyway:
> 
> ```
> mask = np.array([[False], [False],[False],[ True]])
> result = arr[~mask]
> ```

That explanation is wrong- because there is no where I am specifying to delete all elements in any array
